Applications/LLaMA/jni/main.cpp

```diff
@@ -99,80 +100,82 @@ static std::string withKey(const std::string &key,
-  // return argmax if temperature is 0
-  if (temperature < 1e-5) {
-    int argmax_idx =
-      std::distance(logits, std::max_element(logits, logits + NUM_VOCAB));
-    return argmax_idx;
+/**
+ * @brief Apply temperature & top-k & top-p to logits
+ * @return Max logit for softmax
+ */
+float applyTKP(float* logits, int len, float temperature, int top_k, float top_p) {
+  // Apply temperature & Sort logits
+  std::vector<std::pair<int, float>> top_indices_and_logits;
+  for (int i = 0; i < len; ++i) {
+    if (temperature > 1e-5) logits[i] = logits[i] / temperature;
+    top_indices_and_logits.push_back({i, logits[i]});
   }
-
-  // transform logits to softmax
-  std::vector<float> logits_vec;
-  logits_vec.reserve(NUM_VOCAB);
-
-  float max_logits = *std::max_element(logits, logits + NUM_VOCAB);
-  float sum_exp_logits = 0;
-  for (unsigned int i = 0; i < NUM_VOCAB; i++) {
-    float exp_x = exp(logits[i] - max_logits);
-    sum_exp_logits += exp_x;
-    logits_vec.push_back(exp_x);
+  sort(top_indices_and_logits.begin(), top_indices_and_logits.end(), [](auto& a, auto& b){return a.second > b.second; });
+
+  // Accumulate logits
+  float cum_prob = 0;
+  int top_index = 0;
+  while (cum_prob <= top_p && top_index < top_k) {
+    cum_prob += top_indices_and_logits[top_index].second;
+    ++top_index;
   }
 
-  for (unsigned int i = 0; i < NUM_VOCAB; ++i) {
-    logits_vec[i] /= sum_exp_logits;
+  // Apply Top-K and Top-P
+  for (; top_index < len; ++top_index) {
+    logits[top_indices_and_logits[top_index].first] = -INFINITY;
   }
 
-  // sort logits by descending order
-  std::vector<size_t> logits_idx(NUM_VOCAB);
-  std::iota(logits_idx.begin(), logits_idx.end(), 0);
-  std::sort(logits_idx.begin(), logits_idx.end(),
-            [&logits_vec](size_t i1, size_t i2) {
-              return logits_vec[i1] > logits_vec[i2];
-            });
-  std::sort(logits_vec.begin(), logits_vec.end(), std::greater<float>());
-
-  // calculate cumulative logit
-  float cum_logit = 0.0;
-  std::vector<float> cum_logits(NUM_VOCAB);
-  for (size_t i = 0; i < NUM_VOCAB; ++i) {
-    cum_logit += logits_vec[i];
-    cum_logits[i] = cum_logit;
-  }
+  return top_indices_and_logits[0].second;
+}
 
-  // filter logits by temperature
-  size_t mask_idx = 0;
-  for (; mask_idx < NUM_VOCAB; ++mask_idx) {
-    if (cum_logits[mask_idx] > temperature) {
-      break;
+std::vector<int> generate(float *logits,
+             bool do_sample = false,
+             float temperature = 1,
+             int top_k = 1,
+             float top_p = 0,
+             unsigned int NUM_BATCH = 1,
+             unsigned int NUM_VOCAB = 96000) {
+  std::vector<int> outputs;
+  for (unsigned int iteration = 0; iteration < NUM_BATCH; ++iteration) {
+    // return argmax if do_sample is false
+    if (do_sample == false) {
+        int argmax_idx =
+        std::distance(logits, std::max_element(logits, logits + NUM_VOCAB));
+        outputs.push_back(argmax_idx);
```

Comment:
quick question. are we assuming `do_sample == false` equals to `temperature == 0` since it's applying temperature ?

Reply:
@djeong20  `do_sample == false` and `temperature == 0` is different. The do_sample parameter determines whether to apply temperature, top-k, or top-p.

Since it was implemented incorrectly in a previous commit, I made corrections in this one. Thank you for your incisive question!

Reply:
makes sense :) thanks for the clarification!